pkg/solana/logpoller/loader.go

```diff
@@ -0,0 +1,265 @@
+package logpoller
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/gagliardetto/solana-go"
+	"github.com/gagliardetto/solana-go/rpc"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+)
+
+type Parser interface {
+	// ProcessEvent should take a ProgramEvent and parse it based on log signature
+	// and expected encoding. Only return errors that cannot be handled and
+	// should exit further transaction processing on the running thread.
+	//
+	// ProcessEvent should be thread safe.
+	ProcessEvent(ProgramEvent) error
+}
+
+type RPCClient interface {
+	GetLatestBlockhash(ctx context.Context, commitment rpc.CommitmentType) (out *rpc.GetLatestBlockhashResult, err error)
+	GetBlocks(ctx context.Context, startSlot uint64, endSlot *uint64, commitment rpc.CommitmentType) (out rpc.BlocksResult, err error)
+	GetBlockWithOpts(context.Context, uint64, *rpc.GetBlockOpts) (*rpc.GetBlockResult, error)
+	GetSignaturesForAddressWithOpts(context.Context, solana.PublicKey, *rpc.GetSignaturesForAddressOpts) ([]*rpc.TransactionSignature, error)
+	GetTransaction(context.Context, solana.Signature, *rpc.GetTransactionOpts) (*rpc.GetTransactionResult, error)
```

Comment:
Did we end up not needing `GetTransaction`?  I was thinking we would fetch each of the transactions for a particular contract individually, but I guess the way you're doing it here might be more efficient in most cases.

Seems like a good place to start I guess.

Reply:
I started by reading all transactions but found that we SHOULD be able to get the transaction detail from the blocks themselves. This method, if it works for all RPCs, will _drastically_ reduce the number of RPC calls.

Reply:
Yeah, I think it should work for all RPC's. I'm not sure how much it will reduce the number of rpc calls though, because usually each contract wouldn't emit more than 1 event per block (that would be a very high rate of event emission!)... so as long as there are only a handful of contracts involved you'd only rarely get a collision where we needed more than one transaction from a block.  If there are a large number of contracts the node is handling, I could see it helping a lot... but if it's only a small number it would probably be about the same number of rpc calls... but you'd be requesting a lot of extra data, so the network bandwidth might be wasted.

Reply:
I think what I'd lean toward is using one method for "backfills" and a different one for normal polling operation (where the node is already caught up, and there's only been one or a couple new blocks since the last polling tick).

One type of backfill we need is for a single filter when it gets initially registered.  The filter will include a `startingBlock` field to tell it how far to go back. If that is very far back, it could be a large number of events... but we only need them for this one contract. In that case, I think it's pretty clear that we should request them one transaction at a time. It would mean the difference between requesting say, a thousand transactions, and requesting a thousand blocks each with hundreds of transactions. I'd rather we just request 1000 tx's than 100,000 tx's even if the number of rpc calls is the same or slightly more for the 1000. What do you think?

For the other type of backfill, where the node is actually behind and it's trying to catch up on everything... it seems less clear, but I'd probably still lean toward doing it 1 tx at a time. We could also make an arbitrary decision based on how many filters are registered.